CONTRIBUTING.md

```diff
@@ -58,6 +58,262 @@ To contribute a change proposal, use the following workflow:
     5. After you receive feedback from a reviewer, make the requested changes, commit them to your branch, and push them to your remote fork again.
     6. Once approval is given, feel free to squash & merge!
 
+## Writing tests
+
+We use table-driven tests because they allow us to test similar logic on many different cases in a way that is easy to both implement and understand. [This article](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) does a fantastic job explaining the motivation and structure of table-driven testing.
+
+Making table-driven tests in an environment built on the Cosmos SDK has some quirks to it, but overall the structure should be quite similar to what is laid out in the article linked above. We'll lay out two examples below (one that uses our format for messages and another that applies to keeper methods), each of which will hopefully be simple enough to copy-paste into a test file and use as a starting point for your test-writing in the Osmosis Core repo.
+
+### Rules of thumb for table-driven tests:
+
+1. Each test case should test one thing
+2. Each test case should be independent from one another (i.e. ideally, reordering the tests shouldn't cause them to fail)
+3. Functions should not be set as fields for a test case (this usually means that the table-driven approach is being sidestepped and that the logic in the function should probably be factored out to cover multiple/all test cases)
+
```

Comment:
This suggestion is related to: https://github.com/osmosis-labs/osmosis/issues/1947

```suggestion
4. Avoid verbosity in tests by initializing local variables instead of dereferencing struct fields. The antipattern related to this rule is particularly evident with dereferencing `suite.App` to get keepers.
```

We can give [this](https://github.com/osmosis-labs/osmosis/blob/f0270d04bd77cc5e1c23f7913118b3c2ba737e97/x/lockup/keeper/store_test.go#L8-L12) as an example that should be changed to:

```
lockupKeeper := suite.App.LockupKeeper
_ = lockupKeeper.AddLockRefByKey(suite.Ctx, key1, 1)
_ = lockupKeeper.AddLockRefByKey(suite.Ctx, key2, 1)
_ = lockupKeeper.AddLockRefByKey(suite.Ctx, key1, 2)
_ = lockupKeeper.AddLockRefByKey(suite.Ctx, key2, 2)
_ = lockupKeeper.AddLockRefByKey(suite.Ctx, key2, 3)
```

Reply:
This seems like a great thing to me